tools/make_cjk_ranges.swift

```diff
@@ -0,0 +1,391 @@
+/// purpose of emphasis delimiters.
+///
+/// A code point is considered CJK if either of these are true:
+/// - The Script property of the code point is Hangul, or
+/// - Both of the following:
+///   - The East Asian Width property of the code point is `W`, `F`, or `H`, and
+///   - The code point is not a fully-qualified emoji.
+///
+/// For the purposes of this script, only single-code-point emoji are considered. Other code points
+/// that can start or end emoji are handled outside of the ranges given by this script.
+func createCjkTable() async throws
+-> (cjkTable: [UInt32: Bool], cjkDisablingEmojis: [CodePointRange])
+{
+    var cjkTable: [UInt32: Bool] = [:]
+    var cjkDisablingEmojis: [CodePointRange] = []
+
+    // First populate the map with the code points matching the desired East Asian Width values.
+    let (cjkRanges, nonCjkRanges) = try await cjkEawRanges()
+    cjkRanges.forEachCodePoint {
+        cjkTable[$0] = true
+    }
+    nonCjkRanges.forEachCodePoint {
+        cjkTable[$0] = false
+    }
+
+    // Remove code points that can be used as emoji
+    try await singleCodePointEmojiRanges().forEachCodePoint { codePoint in
+        if cjkTable[codePoint] == true {
+            cjkDisablingEmojis.append(.single(codePoint))
+        }
+        cjkTable[codePoint] = false
+    }
+
+    // Add in Hangul code points
+    try await hangulRanges().forEachCodePoint {
+        cjkTable[$0] = true
+    }
+
+    // Fill in the ranges that default to an East Asian Width of W
+    unassignedAsCjkRanges.forEachCodePoint { codePoint in
+        if cjkTable[codePoint] == nil {
+            cjkTable[codePoint] = true
+        }
+    }
+
+    return (cjkTable, cjkDisablingEmojis)
+}
+
+extension [UInt32: Bool] {
+    /// Processes this map of code points into an equivalent set of code point ranges for those
+    /// with a `true` value.
+    func codePointRanges() -> [CodePointRange] {
+        guard !isEmpty else { return [] }
+
+        var ranges: [CodePointRange] = []
+        var rangeStart: UInt32? = nil
+
+        let sortedValues = self.keys.sorted()
+        let maxValue = sortedValues.last!
+
+        for codePoint in 0...maxValue {
+            if self[codePoint] == true {
+                if rangeStart == nil {
+                    rangeStart = codePoint
+                }
+            } else if let start = rangeStart {
+                if start == codePoint - 1 {
+                    ranges.append(.single(start))
+                } else {
+                    ranges.append(.range(start...(codePoint - 1)))
+                }
+                rangeStart = nil
+            }
+        }
+
+        if let rangeStart {
+            if rangeStart == maxValue {
+                ranges.append(.single(rangeStart))
+            } else {
+                ranges.append(.range(rangeStart...maxValue))
+            }
+        }
+
+        return ranges
+    }
+}
+
+print("// Generated by make_cjk_ranges.swift")
+print("")
```

Comment:
I think we should clarify the based Unicode version here.

Reply:
That's a good idea! I can make some callouts about what version of data files i pulled.